### PR TITLE
fix: split insert, update, upsert

### LINKED ISSUE #540: Transaction API returns promises that are never resolved
This issue is caused by several factors.

1. Methods on the Transaction API such as `upsert` return a `Promise`.
2. The typescript definitions of the methods says they returns a `Promise`.
3. The typescript definitions of the methods say they accept callbacks.

The incorrect type definitions cause IDEs such as intellij to warn about missing await statements when calling methods such as `upsert` on a transaction.

When combined with the `upsert` method return a Promise that never resolves leads to bugs that are difficult to diagnose. In the debugger it seems you're receiving a perfectly valid promise misdirecting you in to trying to figure out why the method call never completes.

# Typescript definitions
The issue here is both `Datastore`, and `Transaction` both inherit the same underlying class, `DatastoreRequest`.

This means that typescript will accept method calls to transaction methods such as
```
transaction.upsert(entity, (err) => ...)
await transaction.upsert();
```

This also means the following is considered valid even though it won't actually work;

```
async function execute(writer: DatastoreRequest, item: Item) {
   const key = writer.key([KIND, item.id]);
   const data = toEntity(item);
   await writer.upsert({ key, data });
}
```

This method will work when passed an instance of `Datastore`. However it fails when passed an instance of `Transaction`. This highlights two issues in the type definitions.

1. `Datastore` and `Transaction` are not isomorphic because the type signatures for the methods are not identical (`Transaction` should not include a callback / return a promise).

2. `Transaction` does not contain the `key` method. This should be a compile error, however the definition for `DatastoreRequest` includes `[key: string]: Entity;`, and `Entity` is defined as `any`.

# Transaction methods return a Promise
This is confusing because the Transaction methods are not intended to support callbacks. However due to the underlying `DatastoreRequest` wrapping the methods with `promisifyAll` a promise is returned.

Is leads to a call chain that is effectively:
```
class DatastoreRequest {
    upsert(entities) {
        return new Promise((resolve, reject) => {
            ...
            this.save(entities, (err, result) => {
                if (err) {
                    reject(err);
                } else {
                    resolve(result);
                }
            });
        }
    }
}

class Transaction extends DatastoreRequest {
    save(entities) {
        // callback is ignored, so the promise is never resolved
    }
}
```

When calling this method such as:
```
const tx = datastore.transaction();
await tx.run();
tx.upsert(entity);
```
IDEs such as intellij will warn the user that they have forgotten to handle the Promise returned by the upsert statement. This encourages the user to add an `await`.

This issue is hard to debug though, because when debugging, the call to `upsert` does actually return a Promise, as shown by:

```
const operation = tx.upsert(entity);
console.log(operation); // displays Promise { <unresolved> }
```

I would expect `operation` to be `undefined` in that case as upsert should be a `void` function when called on a transaction.

Lastly, due to the type differences between `Datastore` and `Transaction` this should be an error:
```
function execute(writer: DatastoreRequest) {
    // this should not be possible because update has a
    // different type signature between Datastore and Transaction.
    writer.update(...);
}

const datastore = new Datastore();
const transaction = datastore.transaction();

// this should not be possible because Datastore and Transaction are not compatible
execute(datastore);
execute(transaction);
```

# Related Issues

This issue has been mentioned by previous tickets.

* [#1120 saving basic transaction hangs](https://github.com/googleapis/google-cloud-node/issues/1120)
* [#1351 datastore transaction callbacks hang](https://github.com/googleapis/google-cloud-node/issues/1351)

However, none of those addressed the issue that the combination of the methods returning a promise, and the incorrect method signatures will mislead the user in to using the API incorrectly.

### PR DESCRIPTION
**Pull request by AVaksman**
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #540

### CODE REVIEW

**Review by BenWhitehead — COMMENTED** [review]
In concept I think this is a great idea to clean up the confusion outlined in #540.

I however do not know typescript well enough to evaluate the actual code change, so I'll defer code approval to them.

**Comment by AVaksman** [comment]
Currently functions `insert`, `update`, `upsert` and `save` are implemented in the base class `DatastoreRequest` as asynchronous functions. The functions are then inherited by `Datastore` and `Transaction` classes where `Datastore` functions are meant to be asynchronous, and execute an RPC each, but `Transaction` functions not. The `Transaction` functions are designed to only prep the `entity` and push to `modifiedEntities_` array while awaiting `Transaction.commit()`.
In this PR the aforementioned functions are moved from `DatastoreRequest` class to `Datastore and `Transaction` classes with the intended implementation in each of them.

**Comment by AVaksman** [comment]
Gentle ping

**Review by stephenplusplus — CHANGES_REQUESTED** [review]

File: src/index.ts [hunk 1]
```diff
@@ -425,6 +452,476 @@ class Datastore extends DatastoreRequest {
...
+   *   console.log(key.path); // ['Company', 'donutshack']
+   *   console.log(key.namespace); // 'my-namespace'
+   * });
+   *
+   * //-
+   * // Save different types of data, including ints, doubles, dates, booleans,
+   * // blobs, and lists.
+   * //
+   * // Notice that we are providing an incomplete key. After saving, the
+   * original
```
[the remark this answers is not in the document]
Comment (AVaksman) [reply src/index.ts]: Addressed, thanks.

File: src/index.ts [hunk 2]
```diff
@@ -425,6 +452,476 @@ class Datastore extends DatastoreRequest {
...
+   * datastore.save(entity, (err) => {
+   *   console.log(key.path); // [ 'Company', 5669468231434240 ]
+   *   console.log(key.namespace); // undefined
+   * });
+   *
+   * //-
+   * // Save a single entity using a provided name instead of auto-generated ID.
+   * //
+   * // Here we are providing a key with name instead of an ID. After saving,
+   * the
```
[the remark this answers is not in the document]
Comment (AVaksman) [reply src/index.ts]: Addressed, thanks.

File: src/index.ts [hunk 3]
```diff
@@ -425,6 +452,476 @@ class Datastore extends DatastoreRequest {
...
+   * // original Key object used to save will be updated to contain the path
+   * with
```
[the remark this answers is not in the document]
Comment (AVaksman) [reply src/index.ts]: Addressed, thanks.

File: src/index.ts [hunk 4]
```diff
@@ -425,6 +452,476 @@ class Datastore extends DatastoreRequest {
...
+   *       'Long string (...)',
+   *       {
+   *         description: 'Long string (...)'
+   *       }
+   *     ]
+   *   }
+   * };
+   *
+   * datastore.save(entity, (err, apiResponse) => {});
+   * //-
```
[the remark this answers is not in the document]
Comment (AVaksman) [reply src/index.ts:659]: Added, thanks.

**Review by stephenplusplus — APPROVED** [review]

**Comment by AVaksman** [comment]
Thank you!

**Pull request merged**